## Solution Notebook 

<h3>Notebook Structure </h3>

### Part - I
- Variable, Datastructures
- Loops , Conditional statement
- Functions

### Part - II
- Numpy
- Pandas
- Visualization

# Part-I

### Q1.a Given the string 'great', use an appropriate index command that returns 'a' from the string. 
Given s = ‘great’. 
Expected output: ‘a’

In [ ]:
s = 'great'

In [ ]:
s[3]

'a'

### Q1.b Write a program to find out the second smallest number in the list given below. Expected output:- 5

In [ ]:
li = [32, 3, 5, 8, 33, 7]
li_sorted = sorted(li)
li_sorted

[3, 5, 7, 8, 32, 33]

In [ ]:
li_sorted[1]

5

### Q1.c Find out the unique values of the list given below:
mylist=[1,1,5,3,3,6,11,11]. Expected output: [1,5,3,6,11]


In [ ]:
mylist = [1,1,5,3,3,6,11,11]
set(mylist)

{1, 3, 5, 6, 11}

### Q1.d Print cube root of 216 using appropriate syntax.

In [ ]:
round(216**(1/3))

6

### Q2. Replace the value 'hello' in the below nested list with 'bye'. 
a = [1, 2, [3, 4, 'hello']]
Expected output:  [1, 2, [3, 4, 'bye']]


In [ ]:
a = [1, 2, [3, 4, 'hello']]
a[2]

[3, 4, 'hello']

In [ ]:
a[2][2] = 'bye'
a

[1, 2, [3, 4, 'bye']]

### Q3. Given the below dictionary, write a program that returns [‘object0’].
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}. 
Expected output: [‘object0’]


In [ ]:
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}

In [ ]:
dict['c'][0]

'object0'

### Q4. Write a program to calculate the area of a circle. The program should ask for the appropriate input and print the resulting area on the screen.

Please enter the radius of your circle.

If the user enters a value of 5. The output should be:

The area of the circle is 78.54

This program should indicate that the input is invalid if the user enters a character instead of a number as input. For instance if the user enters a value of ‘a’ instead of 5 in the previous example the program should prompt:

Invalid input, please enter a number: _

In [ ]:
def circle(radius):
    return 3.1415*(radius**2)

In [ ]:
## This is the part of the program that interacts with the user

print('Enter the radius:')
radius  = int(input())
print('The area of the circle is:', circle(radius))

Enter the radius:
5
The area of the circle is: 78.53750000000001


### Q5.  Write a program which will take two lists X & Y as input and it will return a list having all the odd numbers present in the given two lists.
Input list: X=[1,2,3,4,5] & Y=[5,6,7,8,9] 
Expected Output: [1,3,5,5,7,9]


In [ ]:
X=[1,2,3,4,5]
Y=[5,6,7,8,9]

newlist = X+Y
oddlist = []
for i in newlist:
  if i%2!=0:
    oddlist.append(i)

print(oddlist)

[1, 3, 5, 5, 7, 9]


### Q6. Write a function that prints the integers from 1 to 15. But for multiples of three print "GREAT" instead of the number, and for the multiples of five print "LEARNING". For numbers which are multiples of both three and five print "GREATLEARNING".

In [ ]:
def great_learning():
  for i in range(1, 16):
    if i%3 == 0 and i%5==0:
      print("GREATLEARNING")
    elif i%3==0:
      print('GREAT')
    elif i%5==0:
      print("LEARNING")
    else:
      print(i)

great_learning()

1
2
GREAT
4
LEARNING
GREAT
7
8
GREAT
LEARNING
11
GREAT
13
14
GREATLEARNING


# Part-II

### Import necessary libraries and read the provided dataset (online_retail.csv)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df = pd.read_csv('online_sales.csv')

### Check top 5 and random 5 samples of the dataframe. 

In [ ]:
df.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [ ]:
df.sample(5)

InvoiceNo StockCode                          Description  Quantity  \
50707     540551    85180B             PINK HEARTS LIGHT CHAIN          5   
233285    557466     22333    RETROSPOT PARTY BAG + STICKER SET         8   
110673    545692     22326  ROUND SNACK BOXES SET OF4 WOODLAND         48   
29617     538810     22554     PLASTERS IN TIN WOODLAND ANIMALS         1   
54865     540977     22447            PIN CUSHION BABUSHKA BLUE         4   

           InvoiceDate  UnitPrice  CustomerID         Country  
50707     1/10/11 9:43       2.51         NaN  United Kingdom  
233285   6/20/11 13:08       1.65     13815.0         Germany  
110673    3/6/11 12:58       2.55     17428.0  United Kingdom  
29617   12/14/10 12:18       1.65     14871.0  United Kingdom  
54865    1/12/11 15:01       2.51         NaN  United Kingdom

### Check info of the dataframe and write your observations. Comment on data types and shape of the dataset.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 14.6+ MB


- There are 8 columns and 240007 rows
- There are soem object type columns:- InvoiceNo, StockCode, Description, InvoiceDate, Country
- We can convert InvoiceDate, Country to relevant data type
- We can see some null values as well

### Check for null values and report percentage of null values of each column. And drop the rows having null values in it.

In [ ]:
df.isna().sum()

InvoiceNo          0
StockCode          0
Description      901
Quantity           0
InvoiceDate        0
UnitPrice          0
CustomerID     67225
Country            0
dtype: int64

In [ ]:
df.isna().sum()/len(df)*100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.375406
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     28.009600
Country         0.000000
dtype: float64

- 28% of customerID are null

In [ ]:
df = df.dropna()
df.shape

(172782, 8)

### Check statistical summary of the dataset. 

In [ ]:
df.describe()

Quantity      UnitPrice     CustomerID
count  172782.000000  172782.000000  172782.000000
mean       12.244835       3.793566   15274.819941
std       257.898235     101.069930    1725.093177
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13842.000000
50%         5.000000       1.950000   15132.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000

### Drop the instances having quantity less than zero.

In [ ]:
df_new = df[df['Quantity']>=0]
df_new.shape

(168675, 8)

### Check unique values of each country and report the name of the country that has the highest number of instances. 

In [ ]:
df_new['Country'].nunique()

37

In [ ]:
df_new['Country'].value_counts().head(5)

United Kingdom    150046
Germany             3958
France              3562
EIRE                2650
Netherlands         1139
Name: Country, dtype: int64

### Create a new column with name as ‘sales’ having total sales as value. The total sales is defined as Quantity*UnitPrice.

In [ ]:
df_new['sales'] = df_new['Quantity']*df_new['UnitPrice']
df_new.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

    InvoiceDate  UnitPrice  CustomerID         Country  sales  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
5  12/1/10 8:26       7.65     17850.0  United Kingdom  15.30  
6  12/1/10 8:26       4.25     17850.0  United Kingdom  25.50  
7  12/1/10 8:28       1.85     17850.0  United Kingdom  11.10  
8  12/1/10 8:28       1.85     17850.0  United Kingdom  11.10  
9  12/1/10 8:34       1.69     13047.0  United Kingdom  54.08

In [ ]:
df_new.head(1)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   

    InvoiceDate  UnitPrice  CustomerID         Country  sales  
0  12/1/10 8:26       2.55     17850.0  United Kingdom   15.3

### Report the top 5 countries in terms of sales. 
- Consider the size of sales 
- Consider the mean value of sales. 


In [ ]:
df_new.groupby('Country')['sales'].size().sort_values(ascending= False).head(5)

Country
United Kingdom    150046
Germany             3958
France              3562
EIRE                2650
Netherlands         1139
Name: sales, dtype: int64

In [ ]:
df_new.groupby('Country')['sales'].mean().sort_values(ascending= False).head(5)

Country
Australia      126.677444
Netherlands    110.461905
Japan          100.181609
Singapore       90.819912
Sweden          86.532205
Name: sales, dtype: float64

### Report the top 5 products which bring the highest sales. Use StockCode for product information.

In [ ]:
df_new.groupby('StockCode')['sales'].size().sort_values(ascending= False).head(5)

StockCode
85123A    1131
22423      961
47566      787
85099B     704
84879      664
Name: sales, dtype: int64

### Convert the ‘InvoiceDate’ into a dateformate and report the month on which the maximum sales occur?

In [ ]:
# Checking date format
df_new['InvoiceDate'].head(2)

0    12/1/10 8:26
1    12/1/10 8:26
Name: InvoiceDate, dtype: object

In [ ]:
dt = datetime.datetime(2001, 1, 31, 10, 51, 0)

print(dt.strftime('%m/%d/%Y %H:%M'))

01/31/2001 10:51


In [ ]:
df_new['InvoiceDate_new'] = pd.to_datetime(df_new['InvoiceDate'], format='%m/%d/%y %H:%M') 
df_new['InvoiceDate_new'].head(2)

0   2010-12-01 08:26:00
1   2010-12-01 08:26:00
Name: InvoiceDate_new, dtype: datetime64[ns]

In [ ]:
df_new['InvoiceDate_day'] = df_new['InvoiceDate_new'].dt.day

In [ ]:
df_new['InvoiceDate_month'] = df_new['InvoiceDate_new'].dt.month

In [ ]:
df_new.groupby('InvoiceDate_month')['sales'].mean().sort_values(ascending= False)

InvoiceDate_month
1     26.820132
6     23.969489
5     23.959980
2     22.437643
3     21.911939
12    21.892733
4     20.720737
Name: sales, dtype: float64